### ISSUE TITLE
tidy indent+wrap breaks <pre> formatting

### ISSUE THREAD

**Issue by geoffmcl** [htacg/tidy-html5#697]
Issue from https://lists.w3.org/Archives/Public/html-tidy/2018JanMar/0016.html

tidy -indent -wrap  breaks `<pre>` formatting if the line is wrapped

```html
<pre>a long line
</pre>
```

when indented and wrapped ends up like this -

```html
      <pre>
      a long line
</pre>
```

The pre-formatted text shouldn't have leading spaces added.  It should look like

```html
      <pre>
a long line
</pre>
```

example: using tidy --version `HTML Tidy for Cygwin version 5.6.0`

run `$ tidy -indent -wrap 78 --wrap-attributes yes --tidy-mark no`

Input:

```html
<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN"
          "http://www.w3.org/TR/html4/loose.dtd">
<html><head><title>simple test case</title></head><body>
<div><div><div>
<PRE class="SCREEN">x1234567890123456789012345678901234567890123456789
</PRE >
</div></div></div>
</body></html>
```
 
Output:

```html
<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN"
"http://www.w3.org/TR/html4/loose.dtd">
<html>
<head>
  <title>simple test case</title>
</head>
<body>
  <div>
    <div>
      <div>
        <pre class="SCREEN">
        x1234567890123456789012345678901234567890123456789
</pre>
      </div>
    </div>
  </div>
</body>
</html>
```

Regards,
Lee

**Comment by geoffmcl** [comment]
Have pushed this small patch to an `issue-697` branch for consideration and discussion...

This patch has several other consequences!

Yes, it does **avoid** adding any indent to the first text line of the `<pre>`, which is good...

But it also **removes** the indent from the previously indented `<pre>` tag, **and** `<textarea>` tags. Now this **may** be acceptable, **but** it does mean adjusting about 10 of our [regression](https://github.com/htacg/tidy-html5-tests) tests expects output. Again this **may** be acceptable...

And it does **not** remove a newline added after the `<pre>` if the line length is greater than the current wrap value, as discussed in #158...

So at best this is a partial **fix**, with other consequences... so as mentioned maybe there is a better way to address this, and its consequences...

Look forward to further feedback... thanks...

**Comment by geoffmcl** [comment]
As stated was **not** happy with the fix in the `issue-697` branch... it had too many other consequences...

So went back to the drawing board! Started with `next` again...

Have have now created an `issue-697-1` branch, and this is **looking good**! See commit e36df3b for the small diff...

Added 4 test files to my test repo... each should be run with the `-i` indent option. The lines are already greater than the default wrap width of `68`...

 1. The original test [in_697.html](https://github.com/geoffmcl/tidy-test/blob/master/test/input5/in_697.html) now keeps the long line on the same line as the indented `<pre ...>` tag. The `class` attribute is wrapped, but this is **not** a problem.
 
 2. Test [in_697-1.html](https://github.com/geoffmcl/tidy-test/blob/master/test/input5/in_697-1.html) has the closing `</pre>` on the same line, and this is kept in the output. And it has a really really **long** line of words in a `<pre>. . .</pre>` and this is kept in the output.
 
 3. Test [in_697-2.html](https://github.com/geoffmcl/tidy-test/blob/master/test/input5/in_697-2.html) has 3 long lines, and no indent is added to any of them.
 
 4. Test [in_697-3.html](https://github.com/geoffmcl/tidy-test/blob/master/test/input5/in_697-3.html) has a paragraph `<p>` before and after the inline `<pre>` and they are all indented correctly.
 
And the **best** news of all, this `697-1` version of tidy breaks **nothing** in the [regression](https://github.com/htacg/tidy-html5-tests) tests... passes `100%`...

Will continue testing, and will get around to **deleting** the first `issue-697` branch, and creating a PR for this new `issue-697-1` branch...

Would ask others to checkout and test this `issue-697-1` branch and report... thanks...

But presently think, **hope**, this pretty print **bug** is solved...

**Comment by ler762** [comment]
issue-697-1 branch looks good

I did a quick check & haven't seen any cases where it breaks pre formatting.

The bug seems to be solved, so I guess this is a feature request
when indenting, if the `<pre>` block of text is going to take more than one line, have the tag on one line and the text starting on the next line.

https://www.w3.org/TR/html/grouping-content.html#the-pre-element
NOTE:
In the HTML syntax, a leading newline character immediately following the pre element start tag is stripped. 

so
 ```
   <pre>line 1
line 2</pre>
```
is equivalent to
```
    <pre>
line 1
line 2</pre>
```

but much more readable (which is why I'm using indent + wrap)

In any case - thanks for fixing this!

**Comment by geoffmcl** [comment]
@ler762 thanks for testing this in the `issue-697-1` branch, and reporting...

Have now added PR #708 to merge and hopefully close this sometime soon...

As to your `preference` to add a `newline` after the indented `<pre>`, please note, previously, back in tidy circa 2000, when this **was** being done, it also **removed** the indent from the `<pre>`, as in -

```html
  <body>
    <div>
      <div>
        <div>
<pre class="SCREEN">
x1234567890123456789012345678901234567890123456789
</pre>
        </div>
      </div>
    </div>
  </body>
```  

Just saying that it may not be possible to get **both**... but maybe someone can find a way to have **both** ;=))

Anyway if you feel **strongly** about this then please consider opening a **new** separate `Feature Request` for this... thanks...

**Comment by geoffmcl** [comment]
@ler762 have now merged #708, so closing this... thanks...
